### ISSUE TITLE
Tremor aggregation function "stats::dds" produces incorrect values when count is zero 

### ISSUE THREAD

**Issue by anuprout** [tremor-rs/tremor-runtime#546]
**Problem**

It is observed that the aggregate function "stats::dds" produces  fixed numbers for "mean & "max" instead of 0 when the events count is 0 for a certain window. Below is one sample output . 

`puppetlabs.memory,dc=bo1,host_name=localhost,metric_name=non-heap,count_value=0i,max_value=1.7976931348623157e308,mean_value=-2.696539702293474e308,min_value=0.0,p90_value=0.0,p95_value=0.0,p99_value=0.0 1603746984000000000`

As it can be noticed in the above sample output, the **mean & max values are calculated as -2.696539702293474e308 & 1.7976931348623157e308 respectively.** 

Below is the Trickle script used to get this output.

```
use std::type;
use std::record;

define script add_kafka_key

script
use std::json;
let $kafka_key = event.measurement+","+json::encode(event.tags);
emit;
end;

create script add_kafka_key;

define tumbling window `10secs`
with
interval = 10000
end;

create stream normalize;

create stream aggregate;

define generic::batch operator batch
with
count = 3000,
timeout = 5
end;
create operator batch;

select {
"measurement": event.measurement,
"tags": event.tags,
"field": group[2],
"value": event.fields[group[2]],
"timestamp": event.timestamp,
}
from in
group by set(event.measurement, event.tags, each(record::keys(event.fields)))
into aggregate
having type::is_number(event.value);

select
{
"measurement": event.measurement,
"tags": event.tags,
"stats": aggr::stats::dds(event.value, [ "0.9", "0.95", "0.99" ]),
"field": event.field,
"timestamp": aggr::win::first(event.timestamp), # we can't use min since it's a float
}
from aggregate[`10secs`]
group by set(event.measurement, event.tags, event.field)
into normalize;

select {
"measurement": event.measurement,
"tags": event.tags,
"fields": {
"count_{event.field}": event.stats.count,
"min_{event.field}": event.stats.min,
"max_{event.field}": event.stats.max,
"mean_{event.field}": event.stats.mean,
"p90_{event.field}": event.stats.percentiles["0.9"],
"p95_{event.field}": event.stats.percentiles["0.95"],
"p99_{event.field}": event.stats.percentiles["0.99"]
},
"timestamp": event.timestamp,
}
from normalize
into add_kafka_key;

select event from add_kafka_key into batch;
select event from batch into out;
```

**Steps**

1.
2.
3.

**Possible Solution(s)**

**Notes**

Output of `rustup --version`:
Output of `rustup show`:
Output of `tremor --version`:

**Comment by darach** [comment]
Well spotted. On emit the operator should check the count as iff count is zero, set both to 0 or perhaps null as 0 won't always be desirable - in that case count can be checked and an alternate default set in user defined logic.

The hdr implementation has wider stats coverage and the underlying library is hardened better

**Comment by anuprout** [comment]
@darach We were using hdr earlier. But we found an issue with hdr where it is always producing the aggregated data as integer  even though the input data are float. dds seems to be producing in the correct data format. I was going to create a separate  bug for hdr for teh data type issue.

**Comment by darach** [comment]
We can look into enhancing hdr too for float 👍

**Comment by Licenser** [comment]
I want to throw in a few options to handle this depending on what the desired behavior is:

1) using `having event.stats.count > 0` which would remove any measurements that do have a count of 0
2) using a `match` expresson to set the values you want like
```tremor
select
  match event of
    case %{"stats" %{"count" == 0}} =>
      let event.stats.max = 0,
      let event.stats.mean = 0,
      event
    default => event
  end
from ...
```

(disclaimer: non tested pseudo code, not copy&pasta code)

**Comment by anuprout** [comment]
thanks @Licenser . We are doing that in the Trickle script now to drop out the events with  count=0.

**Comment by mfelsche** [comment]
This has been fixed with #1246
